paddle/strings/stringpiece.cc

```diff
@@ -0,0 +1,127 @@
+/*
+  Copyright (c) 2016 PaddlePaddle Authors. All Rights Reserve.
+
+  Licensed under the Apache License, Version 2.0 (the "License");
+  you may not use this file except in compliance with the License.
+  You may obtain a copy of the License at
+
+  http://www.apache.org/licenses/LICENSE-2.0
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+*/
+
+#include "paddle/strings/stringpiece.h"
+
+#include <stddef.h>
+#include <algorithm>
+#include <iosfwd>
+
+namespace paddle {
+
+StringPiece::StringPiece() : data_(NULL), size_(0) {}
+
+StringPiece::StringPiece(const char* d, size_t n) : data_(d), size_(n) {
+  if (d == NULL && n != 0)
+    throw std::invalid_argument(
+        "StringPiece requires len to be 0 for NULL data");
+}
+
+StringPiece::StringPiece(const char* s) : data_(s) {
+  size_ = (s == NULL) ? 0 : strlen(s);
```

Comment:
Oh. I tried it and it seems that we should leave it there; otherwise, we would have to write the following test assertion 

``cpp
EXPECT_EQ("app", TrimSuffix(StringPiece("apple"), "pl"));
```

in a much more complicated way:

```cpp
EXPECT_EQ(std::string("app"), 
     TrimSuffix(StringPiece(std::string("apple")), StringPiece(std::string("pl")));
```